47-backend-frontend/proposal.md

```diff
@@ -0,0 +1,65 @@
+# Summary
+
+Let's create a way to distribute the various procedures that the web node runs across multiple unix processes.
+
+# Motivation
+
+## A story of observability pain
+
+Recently I was required to debug a Concourse cluster, in production, that was having essentially a total outage: The web nodes were no longer serving API requests. They were completely maxing out their memory consumption to the point that they failed their health checks and got restarted by monit (this was a bosh deployment).
+
+I was on a long phone call with some stressed-out operators and everybody was somewhat-aimlessly brainstorming. At one point, somebody said
+
+> it looks like there's a regular pattern of steps - the memory shoots up, stays at max for a minute or so, falls down to zero and then shoots up within about another minute. This doesn't look like a leak, but more like a resource-consumption bug in some kind of periodic asynchronous procedure.
+
+She could definitely be right. In my heightened state (I was trying to fix a production issue in real-time over a Zoom call) I started rapidly reading through the codebase - what are all the periodic asynchronous procedures that the web nodes run?
+
+I was looking through `atc/atccmd/command.go` and realized that the things I was looking for ("periodic asynchronous procedures") all fit more precisely into the category of `github.com/tedsuo/ifrit/grouper.Member` values.
+
+https://github.com/concourse/concourse/blob/release/5.5.x/atc/atccmd/command.go#L796-L890
+
+I started frantically tabulating each of these `grouper.Member`s and the intervals on which they ran, hoping to find something that might possibly cohere with my colleague's observation. Unfortunately it came to nothing, but the plot thickens.
+
+Eventually we got some heap dumps using the pprof endpoint:
+
+https://github.com/concourse/concourse/blob/release/5.5.x/atc/atccmd/command.go#L11
+
+and then we discovered that the `ListAllJobs` endpoint was ultimately mostly responsible for the memory consumption, and we made a PR adding a config flag that to disable specifically that badly-behaved endpoint: https://github.com/concourse/concourse/pull/5340
+
+## A request for non-invasive operator-configurable observability
+
+How can we improve situations like this in the future? I would like the following acceptance criteria:
+
+A way to report compute-resource consumption of Concourse by "member" (or smaller), such that
+* data can be harvested in real-time
```

Comment:
Could you expand on what did you mean by this? Do you mean that that should be continuously collected throughout the execution of the member, or that at the moment I need, I should be able to retrieve it (i.e., take a snapshot)?